drivers/serial/Kconfig.ns16550

```diff
@@ -20,23 +29,41 @@ config UART_NS16550_LINE_CTRL
 
 config UART_NS16550_DRV_CMD
 	bool "Enable Driver Commands"
-	depends on UART_DRV_CMD && UART_NS16550
+	depends on UART_DRV_CMD
 	help
 	  This enables the API for apps to send commands to driver.
 
 	  Says n if not sure.
 
 config UART_NS16750
 	bool "Enable UART 16750 (64-bytes FIFO and auto flow control)"
-	depends on UART_NS16550
 	help
 	  This enables support for 64-bytes FIFO and automatic hardware
 	  flow control if UART controller is 16750.
 
 config UART_NS16550_ACCESS_WORD_ONLY
 	bool "NS16550 only allows word access"
-	depends on UART_NS16550
 	help
 	  In some case, e.g. ARC HS Development kit, the peripheral space of ns
 	  16550 (DesignWare UART) only allows word access, byte access will raise
 	  exception.
+
+config UART_NS16550_PCP
+	bool "Enable support for PRV_CLOCK_PARAMS register"
+	help
+	  This option is for UARTs offering "high-speed" capability through
+	  a register called PRV_CLOCK_PARAMS (aka the PCP).
+
+config UART_NS16550_DLF
+	bool "Enable support for DLF register"
+	help
+	  This option enables support UARTs with the divisor latch fraction
+	  (DLF) capability.
+
+config UART_NS16550_REG_INTERVAL
```

Comment:
how is the config option different than if `reg-shift` exists?

Reply:
I'm not sure I understand the question. This option makes it possible to use `reg-shift` in DT. If this Kconfig option is disabled any `reg-shift` entries are ignored (i.e. it's the responsibility of the SoC to enable this if it knows that such nodes are present). We could remove this (and others as well) Kconfig options if there was a way to check with a DT macro for the existence of a given property (in this case `reg-shift` on any of the matching nodes, but AFAIK such a feature doesn't currently exist.

Reply:
>  We could remove this Kconfig option (and others as well) if there was a way to check with a DT macro for the existence of a given property (in this case `reg-shift`) on any of the matching nodes, but AFAIK such a feature doesn't currently exist.

Where do you need this "way to check"?

In C, there is [DT_NODE_HAS_PROP()](https://docs.zephyrproject.org/latest/reference/devicetree/api.html#c.DT_NODE_HAS_PROP), so no issue there.

Or do you mean you need this in Kconfig? We do have `dt_node_has_bool_prop` in kconfigfunctions.py. It would be trivial to extend this to a function which works with any type of property if that's what is missing.

Reply:
@jhedberg please see also [DT_PROP_OR](https://docs.zephyrproject.org/latest/reference/devicetree/api.html#c.DT_PROP_OR), which may be relevant.

Reply:
@mbolivar the macros you quoted are for when the node_id is known. What I'd need is something that looks for *any* node that matches this driver. The ns16550 driver uses this to decide whether to include certain struct members in the configuration struct for each device instance:
```
 #ifdef CONFIG_UART_NS16550_REG_INTERVAL
          uint8_t reg_interval;
  #endif
```

For PCIe, now that all nodes are appropriately organised into the right topology, there is no need for a Kconfig option since we do this in the struct:
```
  #if DT_ANY_INST_ON_BUS_STATUS_OKAY(pcie)       
          bool pcie;
          pcie_bdf_t pcie_bdf;
          pcie_id_t pcie_id;
  #endif
```

So what I'd need is something like a `DT_ANY_INST_STATUS_OKAY_HAS_PROP(prop_name)` in order to get rid of the added Kconfig options.

The alternative, like I mentioned, would be to accept the increased struct size penalty by always having these members present, but for now I decided to follow the existing design of only including them when necessary.

For reference, previously the following kind of build-time conditions were used for this, however now that the hard-coded assumption of max node instances (used to be 4 for ns16550) has been removed it's no longer possible:
```
#if DT_INST_NODE_HAS_PROP(0, pcp) || \
        DT_INST_NODE_HAS_PROP(1, pcp) || \
        DT_INST_NODE_HAS_PROP(2, pcp) || \
        DT_INST_NODE_HAS_PROP(3, pcp)
#define UART_NS16550_PCP_ENABLED
#endif
```

Reply:
@jhedberg you can build [folds](https://en.wikipedia.org/wiki/Fold_%28higher-order_function%29) using DT_INST_FOREACH_STATUS_OKAY, for example:

```c
#define INST_HAS_FOO_HELPER(inst) DT_INST_NODE_HAS_PROP(inst, foo) ||
#define INST_HAS_BAR_HELPER(inst) DT_INST_NODE_HAS_PROP(inst, bar) ||
#define INST_HAS_BAZ_HELPER(inst) DT_INST_NODE_HAS_PROP(inst, baz) ||

#define ANY_INST_HAS_FOO (DT_INST_FOREACH_STATUS_OKAY(INST_HAS_FOO_HELPER) 0)
#define ANY_INST_HAS_BAR (DT_INST_FOREACH_STATUS_OKAY(INST_HAS_BAR_HELPER) 0)
#define ANY_INST_HAS_BAZ (DT_INST_FOREACH_STATUS_OKAY(INST_HAS_BAZ_HELPER) 0)

#if ANY_INST_HAS_FOO
/* at least one status 'okay' instance has property 'foo' */
#endif
#if ANY_INST_HAS_BAR
/* at least one status 'okay' instance has property 'bar' */
#endif
#if ANY_INST_HAS_BAZ
/* at least one status 'okay' instance has property 'baz' */
#endif
````

Reply:
@mbolivar-nordic excellent, I'll go with that then and remove the Kconfig options. I suspect that might also fix whatever issue @evgeniy-paltsev saw since I probably overlooked some SoC that needed the new options to be enabled.